components/range_cache_memory_engine/src/background.rs

```diff
@@ -604,38 +602,44 @@ impl BackgroundRunnerCore {
     ///
     /// See: [`RangeStatsManager::collect_changes_ranges`] for
     /// algorithm details.
-    fn top_regions_load_evict(&self, delete_range_scheduler: &Scheduler<BackgroundTask>) {
+    async fn top_regions_load_evict(&self, delete_range_scheduler: &Scheduler<BackgroundTask>) {
         let range_stats_manager = match &self.range_stats_manager {
             Some(m) => m,
             None => {
                 return;
             }
         };
-        if range_stats_manager.set_checking_top_regions(true) {
+        if range_stats_manager.ready_for_auto_load_and_evict() {
             return;
         }
 
-        let curr_memory_usage = self.memory_controller.mem_usage();
-        let threshold = self.memory_controller.stop_load_limit_threshold();
-        range_stats_manager.adjust_max_num_regions(curr_memory_usage, threshold);
-
-        let cached_regions = self
-            .engine
-            .region_manager()
-            .regions_map()
-            .read()
-            .cached_regions();
+        let (current_region_count, cached_regions) = {
+            let region_map = self.engine.region_manager().regions_map().read();
+            (region_map.regions().len(), region_map.cached_regions())
+        };
         let (regions_to_load, regions_to_evict) = range_stats_manager
-            .collect_regions_to_load_and_evict(cached_regions, &self.memory_controller);
+            .collect_regions_to_load_and_evict(
+                current_region_count,
+                cached_regions,
+                &self.memory_controller,
+            );
 
-        let mut regions_to_delete = Vec::with_capacity(regions_to_evict.len());
+        let evict_count = regions_to_evict.len();
+        let mut regions_to_delete = Vec::with_capacity(evict_count);
         info!("ime load_evict"; "regions_to_load" => ?&regions_to_load, "regions_to_evict" => ?&regions_to_evict);
+        let (tx, mut rx) = mpsc::channel(evict_count + 1);
         for evict_region in regions_to_evict {
             let cache_region = CacheRegion::from_region(&evict_region);
+            let tx_clone = tx.clone();
+            // Bound is set to 1 so that the sender side will not be blocked
             let deleteable_regions = self.engine.region_manager().evict_region(
                 &cache_region,
                 EvictReason::AutoEvict,
-                None,
+                Some(Box::new(move || {
+                    Box::pin(async move {
```

Comment:
How about using `tokio::sync::Barrier`? In case the callback is somehow dropped and never calls send.

Reply:
When the sender dropped, the receiver gets None and I handled this case below.